src/docs/man/hdb-filter.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "HDB-FILTER 1"
.TH HDB-FILTER 1 "2006-03-28" "Wed Apr  5 13:07:17 EDT 2006" "HDB Documenation"
.SH "NAME"
hdb\-filter \- filter text and replace tags with HDB values 
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
hdb-filter 
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
hdb-filter reads from stdin and replaces tags with \s-1HDB\s0 values. It support the tags \s-1DB\s0, \s-1DBIF\s0, \s-1ELSE\s0, \s-1FI\s0, \s-1SH\s0, \s-1INCLUDE\s0. Tags are specified in the format: 
   \f(CW%:TAG:list\fR.sublist.key%
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Create a tag that will be replaced by the values in list/sublist/key:
.PP
.Vb 1
\&    %:DB:list.sublist.key%
.Ve
.PP
A conditional replace:
.PP
.Vb 5
\&    %:DBIF: list.sublist.key%
\&        some text
\&    %:ELSE:%
\&        some other text
\&    %:FI:%
.Ve
.PP
Run an external script:
.PP
.Vb 1
\&    %:SH:echo hello :%
.Ve
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
hdb-filter returns a zero  exist  status if it succeeds
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
\&\s-1HDB_ROOT\s0 
        A directory path to override the default database directory. Default root is /var/db/hdb
.SH "AUTHOR"
.IX Header "AUTHOR"
Pontus Sodestrom pontus@aptilo.com
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIhdb\-get\fR\|(1) \fIhdb\-del\fR\|(1) \fIhdb\-cut\fR\|(1) \fIhdb\-cut\fR\|(1) \fIhdb\-dump\fR\|(1) \fIhdb\-import\fR\|(1) \fIhdb\-query\fR\|(1) \fIhdb\-find\fR\|(1) \fIhdb\-edit\fR\|(1) 


src/apps/hdb-filter.c
/*************************************************************************
 * HDB, Copyright (C) 2006 Hampus Soderstrom                             *
 * All rights reserved.  Email: hampus@sxp.se                            *
 *                                                                       *
 * This library is free software; you can redistribute it and/or         *
 * modify it under the terms of The GNU Lesser General Public License as *
 * published by the Free Software Foundation; either version 2.1 of the  *
 * License, or (at your option) any later version. The text of the GNU   *
 * Lesser General Public License is included with this library in the    *
 * file LICENSE.TXT.                                                     *
 *                                                                       *
 * This library is distributed in the hope that it will be useful,       *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the files    *
 * LICENSE.TXT for more details.                                         *
 *************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "hdb.h"
#include <unistd.h>

#define CMD_VERSION "$Id: hdb-filter.c,v 1.1 2006/03/12 13:32:13 hampusdb Exp $"

#define TOKEN_LENGTH 1024
#define TOKEN_DELIM  '%'
#if 0 
#define DBG(fmt...) fprintf(stderr, fmt)
#else
#define DBG(fmt...)
#endif

enum state { TOKEN_NONE, TOKEN_START };

static char * next_token(char *s);
static int db_method(char *prefix, char *token, char *postfix);
static int dbif_method(char *prefix, char *token, char *postfix);
static int sh_method(char *prefix, char *token, char *postfix);
static int include_method(char *prefix, char *token, char *postfix);

static int printing = 1;

int main(int argc, char ** argv)
{
  int c;
  char t[TOKEN_LENGTH + 1];
  int i;
  int hdb_net=0;
#ifdef HDB_NET_SWITCH
  int found_lock=0;
#endif
  int state = TOKEN_NONE;
	if(argc > 2){
		printf("hdb-filter V. %s\n", CMD_VERSION);
		exit(1);
	}

   hdb_net=hdb_get_access();
#if HDB_NET_SWITCH
        found_lock=hdb_check_lock(HDB_LOCKFILE);

        //if the daemon isn't started. Pass on the light-hdb
        if(hdb_net == HDB_ACCESS_NETWORK && found_lock){
				DBG("INFO - Switching to light-hdb\n");
                return execvp("lhdb-filter", argv);
        }
        //if the daemon is running. Use it.
        if(hdb_net == HDB_ACCESS_FILE && !found_lock){
                DBG("INFO - Switching to network-hdb\n");
                return execvp("nhdb-filter", argv);
        }
#endif

if(hdb_net){
  if(hdb_connect("127.0.0.1", 12429, NULL, NULL, 0)){
        DBG("ERR - Unable to connect to HDB daemon.\n");
		return 1;
  }	
  if(getenv("HDBROOT")){
	  if(hdb_set_root(getenv("HDBROOT"))){
		  fprintf(stderr, "ERR Unable to set root\n");
		  hdb_disconnect();
		  exit(1);
	  }
  }	
}

  while((c = getchar()) != EOF)
  {
    switch(state)
    {
      case TOKEN_START : 
	if (c == TOKEN_DELIM)
	{
	  int  ret;
	  char *prefix  = NULL;
	  char *method  = NULL;
	  char *token   = NULL;
	  char *postfix = NULL;
	  char *s = strdup(t);
	  char n;
	 
	  prefix  = s;
	  method  = next_token(prefix);
	  token   = next_token(method);
	  postfix = next_token(token);

	  DBG("prefix = %s\n", prefix);
	  DBG("method = %s\n", method);
	  DBG("token = %s\n", token);
	  DBG("postfix = %s\n", postfix);

	  if (token == NULL)
	  {
	    ret = 0;
	  }
	  else if (!strcmp(method, "DB"))
	  {
	    ret = db_method(prefix, token, postfix);
	  }
	  else if (!strcmp(method, "DBIF"))
	  {
	    ret = dbif_method(prefix, token, postfix);
            n = getchar(); if (n != '\n') ungetc(n, stdin);
	  }
	  else if (!strcmp(method, "ELSE"))
	  {
	    ret = 1;
	    printing = !printing;
            n = getchar(); if (n != '\n') ungetc(n, stdin);
	  }
	  else if (!strcmp(method, "FI"))
	  {
	    ret = 1;
	    printing = 1;
            n = getchar(); if (n != '\n') ungetc(n, stdin);
	  }
	  else if (!strcmp(method, "SH"))
	  {
	    ret = sh_method(prefix, token, postfix);
	  }
	  else if (!strcmp(method, "INCLUDE"))
	  {
	    ret = include_method(prefix, token, postfix);
	  }
	  else
	  {
	    ret = 0;
	  }
	  
	  if (!ret)
	  {
	    if (printing) printf("%c%s%c", TOKEN_DELIM, t, TOKEN_DELIM);
	  }
	  free(s);
	  state = TOKEN_NONE;	  	  
	}
	else if (c == '\n' || i == TOKEN_LENGTH)
	{
	  state = TOKEN_NONE;
	  if (printing) printf("%c%s%c", TOKEN_DELIM, t, c);
	}
	else
	{
	  t[i++] = c;	  
	}
	break;
      case TOKEN_NONE :
	if (c == TOKEN_DELIM)
	{
	  state = TOKEN_START;
	  i = 0;
	  memset(t, 0, TOKEN_LENGTH);
	}
	else
	{
	  if (printing) putchar(c);
	}
	break; 
    default: break;
    }
  }
	if(hdb_net){
		hdb_disconnect();
	}
  return 0;
}

static char* next_token(char *s)
{
  if (s == NULL)
  {
    return s;
  }

  while(*s)
  {
    if (*s == '\\')
    {
      s++;
      if (*s) s++;
      continue;
    }

    if (*s == ':')
    {
      *s = '\0';
      return (s+1);
    }
    s++;
  }
  return NULL;
}

static int db_method(char *prefix, char *token, char *postfix)
{	        
  char list[256];
  char *db_key;
  char *val;
  int i;

  if (!printing)
    return 1;

  db_key = strrchr(token, '.');
  DBG("db_key = %s\n", db_key);
  if (db_key)
  {
    *db_key = 0; 
    db_key++;
  }
  i = 0;
  while (*token && i < 256)
  {
    if (*token == '.')
    {
      *token = '/';
    }
    else if (*token == '\\')
    {
      token++;
      if (*token == '\0') break;
    }
    list[i] = *token;
    i++;
    token++;
  }
  list[i] = 0;
  val = hdb_get_val(list, db_key);
  DBG("val = %s\n", val);
  if (val && strcmp(val, ""))
  {
    printf("%s%s%s", 
	   prefix  ? prefix  : "",
	   val     ? val     : "",
	   postfix ? postfix : "");
    free(val);
  }
  return 1;
}

static int dbif_method(char *prefix, char *token, char *postfix)
{	        
  char list[256];
  char *db_key;
  int  val;
  int i;

  db_key = strrchr(token, '.');
  DBG("db_key = %s\n", db_key);
  if (db_key)
  {
    *db_key = 0; 
    db_key++;
  }
  i = 0;
  while (*token && i < 256)
  {
    if (*token == '.')
    {
      *token = '/';
    }
    else if (*token == '\\')
    {
      token++;
      if (*token == '\0') break;
    }
    list[i] = *token;
    i++;
    token++;
  }
  list[i] = 0;
  val = hdb_get_bool(list, db_key);
  DBG("val = %d\n", val);
  if (val)
  {
    printing = 1;
  }
  else
  {
    printing = 0;
  }
  return 1;
}

static int sh_method(char * prefix, char *token, char *postfix)
{
  FILE* p;
  int c;

  if (!printing)
    return 1;

  p = popen(token, "r");
  if (p == NULL)
  {
    return 1;
  }

  c = fgetc(p);
  if (c == EOF)
  {
    pclose(p);
    return 1;
  }

  printf("%s%c", prefix ? prefix : "", c);

  while ((c = fgetc(p)) != EOF)
  {
    if (c == '\\')
    {
      continue;
    }
    putchar(c);
  }
  pclose(p);
  printf("%s", postfix ? postfix : "");
  return 1;
}

static int include_method(char *prefix, char *token, char * postfix)
{
  FILE* p;
  int c;
  
  if (!printing) 
    return 1;

  p = fopen(token, "r");
  if (p == NULL)
  {
    return 1;
  }

  c = fgetc(p);

  if (c == EOF)
  {
    pclose(p);
    return 1;
  }

  printf("%s%c", prefix ? prefix : "", c);

  while ((c = fgetc(p)) != EOF)
  {
    putchar(c);
  }
  pclose(p);
  printf("%s", postfix ? postfix : "");
  return 1;  
}
